File: project_euler/src/problem_039.erl
%% @author std-string

%% If p is the perimeter of a right angle triangle with integral length sides, {a,b,c}, there are exactly three solutions for p = 120.
%% {20,48,52}, {24,45,51}, {30,40,50}
%% For which value of p <= 1000, is the number of solutions maximised?

-module(problem_039).
-export([get_check_data/0, prepare_data/2, solve/1]).

-behaviour(numerical_task_behaviour).

-type triangle() :: {A :: pos_integer(), B :: pos_integer(), C :: pos_integer()}.
-type triangles() :: [triangle()].

%% ====================================================================
%% API functions
%% ====================================================================

-spec get_check_data() -> [{Input :: term(), Output :: term()}].
get_check_data() -> [{1000, 840}].

-spec prepare_data(ModuleSourceDir :: string(), Input :: term()) -> term().
prepare_data(_ModuleSourceDir, Input) -> Input.

-spec solve(PreparedInput :: term()) -> term().
solve(MaxPerimeter) ->
    {DestPerimeter, _DestCount} = lists:foldl(fun(Current, {Perimeter, MaxCount}) ->
        CurrentCount = length(solve_right_triangle(Current)),
        if
            CurrentCount > MaxCount -> {Current, CurrentCount};
            CurrentCount =< MaxCount -> {Perimeter, MaxCount}
        end
    end, {0, 0}, lists:seq(1, MaxPerimeter)),
    DestPerimeter.

%% ====================================================================
%% Internal functions
%% ====================================================================

-spec solve_right_triangle(Perimeter :: pos_integer()) -> triangles().
solve_right_triangle(Perimeter) ->
    [{A, B, C} || A <- lists:seq(1, Perimeter),
                  B <- lists:seq(1, Perimeter - A),
                  C <- [Perimeter - A - B],
                  C > 0,
                  A =< B,
                  A + B > C,
                  A*A + B*B == C*C].

File: projecteuler.fsharp/ProjectEulerTasks/Problem039.fs
namespace ProjectEulerTasks

open NUnit.Framework
open ProjectEulerTasks.Utils
open CommonLib
open System.Collections.Generic

// If p is the perimeter of a right angle triangle with integral length sides, {a,b,c}, there are exactly three solutions for p = 120: {20, 48, 52}, {24, 45, 51}, {30, 40, 50}
// For which value of p <= 1000, is the number of solutions maximised?

[<TestFixture>]
type Problem039() =

    let solveImpl (maxPerimeter: int) =
        let generator = PythagoreanTriplesGenerator(fun triple -> triple.X + triple.Y + triple.Z)
        let triples = generator.GenerateTriples(maxPerimeter)
        let storage = Dictionary<int, int>()
        for triple in triples do
            let perimeter = triple.X + triple.Y + triple.Z
            if not (storage.ContainsKey(perimeter)) then
                storage.[perimeter] <- 0
            storage.[perimeter] <- storage.[perimeter] + 1
        (storage |> Seq.maxBy (fun kvPair -> kvPair.Value)).Key

    [<TestCase(1000, 840, TimeThresholds.HardTimeLimit)>]
    member public this.Solve(maxPerimeter: int, expectedAnswer: int, timeLimit: int) =
        SolutionUtils.CheckSolution(timeLimit, expectedAnswer, solveImpl, maxPerimeter)
